Mathlib/Tactic/DeriveCountable.lean

```diff
@@ -0,0 +1,276 @@
+Authors: Kyle Miller
+-/
+import Lean.Meta.Transform
+import Lean.Meta.Inductive
+import Lean.Elab.Deriving.Basic
+import Lean.Elab.Deriving.Util
+import Mathlib.Data.Countable.Defs
+import Mathlib.Data.Nat.Pairing
+
+/-!
+# `Countable` deriving handler
+
+Adds a deriving handler for the `Countable` class.
+-/
+
+namespace Mathlib.Deriving.Countable
+open Lean Parser.Term Elab Deriving Meta
+
+/-!
+### Theory
+
+We use the `Nat.pair` function to encode an inductive type in the natural numbers,
+following a pattern similar to the tagged s-expressions used in Scheme/Lisp.
+We develop a little theory to make constructing the injectivity functions very straightforward.
+
+This is easiest to explain by example. Given a type
+```lean
+inductive T (α : Type)
+  | a (n : α)
+  | b (n m : α) (t : T α)
+```
+the deriving handler constructs the following three declarations:
+```lean
+noncomputable def T.toNat (α : Type) [Countable α] : T α → ℕ
+  | a n => Nat.pair 0 (Nat.pair (encode n) 0)
+  | b n m t => Nat.pair 1 (Nat.pair (encode n) (Nat.pair (encode m) (Nat.pair t.toNat 0)))
+
+theorem T.toNat_injective (α : Type) [Countable α] : Function.Injective (T.toNat α) := fun
+  | a .., a .. => by
+    refine cons_eq_imp_init ?_
+    refine pair_encode_step ?_; rintro ⟨⟩
+    intro; rfl
+  | a .., b .. => by refine cons_eq_imp ?_; rintro ⟨⟩
+  | b .., a .. => by refine cons_eq_imp ?_; rintro ⟨⟩
+  | b .., b .. => by
+    refine cons_eq_imp_init ?_
+    refine pair_encode_step ?_; rintro ⟨⟩
+    refine pair_encode_step ?_; rintro ⟨⟩
+    refine cons_eq_imp ?_; intro h; cases T.toNat_injective α h
+    intro; rfl
+
+instance (α : Type) [Countable α] : Countable (T α) := ⟨_, T.toNat_injective α⟩
+```
+-/
+
+private noncomputable def encode {α : Sort*} [Countable α] : α → ℕ :=
+  (Countable.exists_injective_nat α).choose
+
+private noncomputable def encode_injective {α : Sort*} [Countable α] :
+    Function.Injective (encode : α → ℕ) :=
+    (Countable.exists_injective_nat α).choose_spec
+
+/--
+Initialize the injectivity argument. Pops the constructor tag.
+-/
+private theorem cons_eq_imp_init {p : Prop} {a b b' : ℕ}
+    (h : b = b' → p) : Nat.pair a b = Nat.pair a b' → p := by
+  simpa [Nat.pair_eq_pair, and_imp] using h
+
+/--
+Generic step of the injectivity argument, pops the head of the pairs.
+Used in the recursive steps where we need to supply an additional injectivity argument.
+-/
+private theorem cons_eq_imp {p : Prop} {a b a' b' : ℕ}
+    (h : a = a' → b = b' → p) : Nat.pair a b = Nat.pair a' b' → p := by
+  rwa [Nat.pair_eq_pair, and_imp]
+
+/--
+Specialized step of the injectivity argument, pops the head of the pairs and decodes.
+-/
+private theorem pair_encode_step {p : Prop} {α : Sort*} [Countable α]
+    {a b : α} {m n : ℕ}
+    (h : a = b → m = n → p) : Nat.pair (encode a) m = Nat.pair (encode b) n → p :=
+  cons_eq_imp fun ha => h (encode_injective ha)
+
+
+/-!
+### Implementation
+-/
+
+/-!
```

Comment:
```suggestion

```
Is there a reason for not merging these module docs?

Reply:
Yes, otherwise it will look like this module doc is explaining the heading, but it's marking a section of the implementation.